fluent/fluent_test.go

```diff
@@ -274,116 +284,220 @@ func TestJsonConfig(t *testing.T) {
+			t.Parallel()
+
+			d := newTestDialer()
+			var f *Fluent
+			defer func() {
+				if f != nil {
+					f.Close()
+				}
+			}()
+
+			go func() {
+				var err error
+				if f, err = newWithDialer(tc, d); err != nil {
+					t.Errorf("Unexpected error: %v", err)
+				}
+
+				_ = f.PostWithTime("tag_name", time.Unix(1482493046, 0), map[string]string{"foo": "bar"})
+				_ = f.PostWithTime("tag_name", time.Unix(1482493050, 0), map[string]string{"fluentd": "is awesome"})
+			}()
+
+			conn := d.waitForNewConn(true)
+			assertReceived(t,
+				conn.waitForNextWrite(true, ""),
+				"[\"acme.tag_name\",1482493046,{\"foo\":\"bar\"},{}]")
+
+			assertReceived(t,
+				conn.waitForNextWrite(true, ""),
+				"[\"acme.tag_name\",1482493050,{\"fluentd\":\"is awesome\"},{}]")
+		})
 	}
-	ch := make(chan result, 1)
+}
+
+func TestReconnectAndResendAfterTransientFailure(t *testing.T) {
+	testcases := map[string]Config{
+		"with Async": {
+			Async:         true,
+			MarshalAsJSON: true,
+		},
+		"without Async": {
+			Async:         false,
+			MarshalAsJSON: true,
+		},
+	}
+
+	for tcname := range testcases {
+		t.Run(tcname, func(t *testing.T) {
+			tc := testcases[tcname]
+			t.Parallel()
+
+			d := newTestDialer()
+			var f *Fluent
+			defer func() {
+				if f != nil {
+					f.Close()
+				}
+			}()
+
+			go func() {
+				var err error
+				if f, err = newWithDialer(tc, d); err != nil {
+					t.Errorf("Unexpected error: %v", err)
+				}
+
+				_ = f.EncodeAndPostData("tag_name", time.Unix(1482493046, 0), map[string]string{"foo": "bar"})
+				_ = f.EncodeAndPostData("tag_name", time.Unix(1482493050, 0), map[string]string{"fluentd": "is awesome"})
+			}()
+
+			// Accept the first connection dialing and write.
+			conn := d.waitForNewConn(true)
+			assertReceived(t,
+				conn.waitForNextWrite(true, ""),
+				"[\"tag_name\",1482493046,{\"foo\":\"bar\"},{}]")
+
+			// The next write will fail and the next connection dialing will be dropped
+			// to test if the logger is reconnecting as expected.
+			conn.waitForNextWrite(false, "")
+			d.waitForNewConn(false)
+
+			// Next, we allow a new connection to be established and we allow the last message to be written.
+			conn = d.waitForNewConn(true)
+			assertReceived(t,
+				conn.waitForNextWrite(true, ""),
+				"[\"tag_name\",1482493050,{\"fluentd\":\"is awesome\"},{}]")
+		})
+	}
+}
+
+func timeout(t *testing.T, duration time.Duration, fn func(), reason string) {
+	done := make(chan struct{})
 	go func() {
-		config := Config{
-			FluentPort: 8888,
-			Async:      true,
-		}
-		f, err := New(config)
-		ch <- result{f: f, err: err}
+		fn()
+		done <- struct{}{}
 	}()
 
 	select {
-	case res := <-ch:
-		if res.err != nil {
-			t.Errorf("fluent.New() failed with %#v", res.err)
-			return
-		}
-		res.f.Close()
-	case <-time.After(time.Millisecond * 500):
-		t.Error("Async must not block")
+	case <-time.After(duration):
+		t.Fatalf("time out after %s: %s", duration.String(), reason)
+	case <-done:
+		return
```

Comment:
Should we check `case <-done` before `case <-time.After()`?
(I'm not sure if the order of code affects the order of processes actually called)

Reply:
By looking at Go spec (see [here](https://golang.org/ref/spec#Select_statements)), I'd say the order doesn't matter since there're no default case and both receive cases here are blocking :wink: